pyomo/core/base/indexed_component.py

```diff
@@ -673,11 +673,15 @@ def _setitem_when_not_present(self, index, value):
     def set_value(self, value):
         """Set the value of a scalar component."""
         if self.is_indexed():
-            raise ValueError(
+            if type(value) is dict:
```

Comment:
Should we specifically look for `dict`, or would it be better / more flexible to do:
```
if hasattr(value, '__getitem__') and hasattr(value, '__iter__'):
    for key in value:
        self[key] = value[key]
```
...or maybe just put the thing in a try-catch block and any AttributeError would get mapped to the the ValueError below (I am less enamoured with that, as we have been bitten in the past for suppressing / replacing an exception?